astropy/io/fits/tests/test_core.py

```diff
@@ -996,6 +998,60 @@ def test_filename_with_colon(self):
         with fits.open(self.temp(filename)) as hdul:
             assert np.all(hdul[0].data == hdu.data)
 
+
+    def test_writeto_full_disk(self, monkeypatch):
+        """
+        Test that it gives a readable error when trying to write an hdulist
+        to a full disk.
+        """
+        def _writeto(self, array):
+            raise IOError("Fake error raised when writing file.")
+
+        def get_free_space_in_dir(path):
+            return 0
+
+        with pytest.raises(IOError) as exc:
+            monkeypatch.setattr(fits.hdu.base._BaseHDU, "_writeto", _writeto)
+            monkeypatch.setattr(data, "get_free_space_in_dir", get_free_space_in_dir)
+
+            n = np.arange(0, 1000)
+            hdu = fits.PrimaryHDU(n)
+            hdulist = fits.HDUList(hdu)
+            filename = self.temp('test.fits')
+            hdulist.writeto(filename)
+            hdulist.close()
+
+            assert "Not enough space on disk. Fake error raised for writing to files" in exc.value
+
+    def test_flush_full_disk(self, monkeypatch):
+        """
+        Test that it gives a readable error when trying to update an hdulist
+        to a full disk.
+        """
+        filename = self.temp('test.fits')
+        hdul = [fits.PrimaryHDU(), fits.ImageHDU()]
+        hdul = fits.HDUList(hdul)
+        hdul[0].data = np.arange(0, 1000)
+        hdul.writeto(filename)
+
+        def _writeto(self, array, inplace=False, copy=False):
+            raise IOError("Fake error raised when writing file.")
+
+        def get_free_space_in_dir(path):
+            return 0
+
+        monkeypatch.setattr(fits.hdu.base._BaseHDU, "_writeto", _writeto)
+        monkeypatch.setattr(data, "get_free_space_in_dir", get_free_space_in_dir)
+
+        with pytest.raises(IOError) as exc:
+            filename = self.temp('test.fits')
+            hdul = fits.open(filename, mode='update')
```

Comment:
This should be opened instead with its own `with` statement (nested under the `with pytest.raises`).  The reason being that the error probably actually occurs with the `hdul.flush()`, I think, so the `hdul.close()` is never actually run (nor, for that matter, is the assert, which should just be outside the `with` statement entirely).